sdk/core/core/src/az_json_data.c

```diff
@@ -0,0 +1,220 @@
+  ((az_data){ \
+      .span = (az_span){ .begin = (uint8_t const *)&(V), .size = sizeof(V) }, \
+      .align = AZ_ALIGNOF(V), \
+  })
+
+typedef struct {
+  size_t size;
+  size_t align;
+} _az_type;
+
+#define _AZ_TYPE(T) \
+  ((_az_type){ \
+      .size = sizeof(T), \
+      .align = AZ_ALIGNOF(T), \
+  })
+
+uint8_t * _az_align_floor(uint8_t * const p, size_t const align) {
+  return (uint8_t *)(((size_t)p) / align * align);
+}
+
+uint8_t * _az_align_ceil(uint8_t * const p, size_t const align) {
+  return _az_align_floor(p + (align - 1), align);
+}
+
+AZ_NODISCARD az_result _az_span_builder_aligned_append(
+    az_span_builder * const builder,
+    az_data const data,
+    void ** const out) {
+  // alignment.
+  {
+    uint8_t * const p = builder->buffer.begin + builder->length;
+    AZ_RETURN_IF_FAILED(az_span_builder_append_zeros(builder, _az_align_ceil(p, data.align) - p));
+  }
+  *out = builder->buffer.begin + builder->length;
+  return az_span_builder_append(builder, data.span);
+}
+
+AZ_NODISCARD az_result
+_az_span_builder_top_aligned_append(az_span_builder * const builder, az_data const data) {
+  size_t const size = builder->buffer.size;
+  if (size < data.span.size) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+  uint8_t * const begin = builder->buffer.begin;
+  uint8_t * const end = _az_align_floor(begin + size - data.span.size, data.align);
+  if (end < begin + builder->length) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+  size_t const new_size = end - begin;
+  {
+    az_mut_span result = { 0 };
+    AZ_RETURN_IF_FAILED(
+        az_mut_span_move(az_mut_span_drop(builder->buffer, new_size), data.span, &result));
+  }
+  builder->buffer.size = new_size;
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result _az_span_builder_top_array_revert(
+    az_span_builder * const builder,
+    _az_type const item_type,
+    size_t const new_size,
+    void ** const out_array_begin,
+    size_t * const out_array_size) {
+  AZ_CONTRACT_ARG_NOT_NULL(builder);
+  AZ_CONTRACT_ARG_NOT_NULL(out_array_begin);
+  AZ_CONTRACT_ARG_NOT_NULL(out_array_size);
+
+  // 3. restore buffer size
+  size_t const offset = builder->buffer.size;
+  builder->buffer.size = new_size;
+  az_mut_span const buffer = builder->buffer;
+
+  // 1. revert the array.
+  size_t const item_count = (new_size - offset) / item_type.size;
+  for (size_t i = 0; i < item_count / 2; ++i) {
+    az_mut_span const a
+        = az_mut_span_take(az_mut_span_drop(buffer, offset + i * item_type.size), item_type.size);
+    az_mut_span const b = az_mut_span_take(
+        az_mut_span_drop(buffer, offset + (item_count - i - 1) * item_type.size), item_type.size);
+    az_mut_span_swap(a, b);
+  }
+
+  // 2. move it to front.
+  az_data const data = {
+    .span
+    = az_span_take(az_span_drop(az_mut_span_to_span(buffer), offset), item_count * item_type.size),
+    .align = item_type.align,
+  };
+  AZ_RETURN_IF_FAILED(_az_span_builder_aligned_append(builder, data, out_array_begin));
```

Comment:
I'd like to see tests for each of these failure conditions, to ensure the buffers are in the states you expect after the failure occurs.